core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala

```diff
@@ -57,53 +57,57 @@ case class CoprocessorRDD(output: Seq[Attribute], tiRdd: TiRDD) extends LeafExec
   val internalRdd: RDD[InternalRow] = RDDConversions.rowToRowRdd(tiRdd, output.map(_.dataType))
   private lazy val project = UnsafeProjection.create(schema)
 
-  protected override def doExecute(): RDD[InternalRow] = {
-    val numOutputRows = longMetric("numOutputRows")
-    // In Spark 2.3.0 the method declaration is:
-    // private[spark] def mapPartitionsWithIndexInternal[U: ClassTag](
-    //      f: (Int, Iterator[T]) => Iterator[U],
-    //      preservesPartitioning: Boolean = false): RDD[U]
-    //
-    // In other Spark 2.3.x versions, the method declaration is:
-    // private[spark] def mapPartitionsWithIndexInternal[U: ClassTag](
-    //      f: (Int, Iterator[T]) => Iterator[U],
-    //      preservesPartitioning: Boolean = false,
-    //      isOrderSensitive: Boolean = false): RDD[U]
-    //
-    // Hereby we use reflection to support different Spark versions.
-    val method = try {
+  // In Spark 2.3.0 the method declaration is:
+  // private[spark] def mapPartitionsWithIndexInternal[U: ClassTag](
+  //      f: (Int, Iterator[T]) => Iterator[U],
+  //      preservesPartitioning: Boolean = false): RDD[U]
+  //
+  // In other Spark 2.3.x versions, the method declaration is:
+  // private[spark] def mapPartitionsWithIndexInternal[U: ClassTag](
+  //      f: (Int, Iterator[T]) => Iterator[U],
+  //      preservesPartitioning: Boolean = false,
+  //      isOrderSensitive: Boolean = false): RDD[U]
+  //
+  // Hereby we use reflection to support different Spark versions.
+  @transient private val method = try {
+    classOf[RDD[InternalRow]]
+      .getDeclaredMethod(
+        "mapPartitionsWithIndexInternal",
+        classOf[(Int, Iterator[InternalRow]) => Iterator[UnsafeRow]],
+        classOf[Boolean],
+        classOf[Boolean],
+        classOf[ClassTag[UnsafeRow]]
+      )
+  } catch {
+    case _: Throwable =>
```

Comment:
1. it's better to provide a general function to do spark version related reflection
2. you can read spark version from `org.apache.spark.SPARK_VERSION`, and then call the correct reflected function. Do not need to use try catch.

Reply:
agree.